employee_api/handlers/employee.go

```diff
@@ -0,0 +1,246 @@
+		log.Fatal("This is the error:", err)
+		return
+	}
+
+	s.DB = db
+
+	s.DB.AutoMigrate(&models.Employee{})
+
+	s.Router = mux.NewRouter()
+	s.Routes = s.collectRoutes()
+	for _, route := range s.Routes {
+		s.Router.HandleFunc(route.Pattern, route.HandlerFunc).Methods(route.Method)
+	}
+}
+
+func (s *Server) create(w http.ResponseWriter, r *http.Request) {
+	// read data from request body
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		responses.Error(w, http.StatusBadRequest, []string{"Invalid request"})
+		return
+	}
+
+	// convert data to Employee struct
+	employee := models.Employee{}
+	err = json.Unmarshal(body, &employee)
+	if err != nil {
+		responses.Error(w, http.StatusBadRequest, []string{"Invalid request"})
+		return
+	}
+
+	// validate data
+	messages := employee.Validate()
+	if len(messages) > 0 {
+		responses.Error(w, http.StatusBadRequest, messages)
+		return
+	}
+
+	tx := s.DB.Create(&employee)
+	if tx.Error != nil {
+		messages = []string{tx.Error.Error()}
+		responses.Error(w, http.StatusBadRequest, messages)
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
+	responses.Success(w, http.StatusCreated,
+		struct {
+			ID int `json:"id"`
+		}{
+			ID: employee.ID,
+		})
+}
+
+func (s *Server) delete(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	idQuery := vars["id"]
+	id, err := strconv.Atoi(idQuery)
+	if err != nil {
+		responses.Error(w, http.StatusBadRequest, []string{"Invalid id"})
+		return
+	}
+
+	tx := s.DB.Delete(&models.Employee{}, id)
+	if tx.Error != nil {
+		messages := []string{tx.Error.Error()}
+		responses.Error(w, http.StatusBadRequest, messages)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	responses.Success(w, http.StatusOK, struct {
+		ID int `json:"id"`
+	}{
+		ID: id,
+	})
+}
+
+func (s *Server) get(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	var employees []models.Employee
+
+	idQuery := vars["id"]
+	// convert string to int
+	id, err := strconv.Atoi(idQuery)
+	if err != nil {
+		responses.Error(w, http.StatusBadRequest, []string{"Invalid id"})
+	}
+
+	results := s.DB.First(&employees, id)
+	if results.Error != nil {
+		responses.Error(w, http.StatusBadRequest, []string{results.Error.Error()})
+		return
+	}
+	responses.Success(w, http.StatusOK, employees)
+}
+
+func (s *Server) update(w http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		responses.Error(w, http.StatusBadRequest, []string{"Invalid request"})
+		return
+	}
+
+	vars := mux.Vars(r)
+	idQuery := vars["id"]
+	id, err := strconv.Atoi(idQuery)
+	if err != nil {
+		responses.Error(w, http.StatusBadRequest, []string{"Invalid request"})
+		return
+	}
+
+	var updateEmployee models.Employee
+	err = json.Unmarshal(body, &updateEmployee)
+	if err != nil {
+		responses.Error(w, http.StatusBadRequest, []string{"Invalid request"})
+		return
+	}
+
+	// update employee
+	tx := s.DB.Model(&models.Employee{}).Where("id = ?", id).Updates(updateEmployee)
```

Comment:
split into storage, we can handle more case errors in here, eg the employee is not found, or db error, to return it correctly to client.

Reply:
I move it to employee_service/storage/employee.go